File: 0202-happy-number/0202-happy-number.cpp
// TC- O(logn)
// SC- O(1)

class Solution {
private:
    int sum;
    int getNext(int n){
        sum = 0;
        while(n){
            sum += ((n%10) * (n%10));
            n = n/10;
            // cout<<n<<" ";
        }
        return sum;
    }
public:
    bool isHappy(int n) {
        int slowptr = n;
        int fastptr = getNext(n);
        while(fastptr!=1 and slowptr!=fastptr){
            slowptr = getNext(slowptr);
            fastptr = getNext(getNext(fastptr));
        }
        return fastptr==1;
    }
};

// // TC- O(logn)
// // SC- O(logn)

// class Solution {
// private:
//     int sum;
//     int getNext(int n){
//         sum = 0;
//         while(n){
//             sum += ((n%10) * (n%10));
//             n = n/10;
//             // cout<<n<<" ";
//         }
//         return sum;
//     }
// public:
//     bool isHappy(int n) {
//         unordered_set<int> mem;
//         while(n!=1){
//             if(mem.count(n)!=0)
//                 return false;
//             mem.insert(n);
//             n = getNext(n);
//             // cout<<n<<" ";
//         }
//         return true;
//     }
// };

File: 0202-happy-number/0202-happy-number.py
# TC- O(logn) - logn for one getNext call, max num of times it can be called for 32 bit integer is 10*9*9 = 810, so O(logn*1)
# SC- O(1)

class Solution:
    def getNext(self, n):
        square_sum = 0
        while(n):
            rem = n%10
            square_sum += rem*rem
            n = n//10
        return square_sum

    def isHappy(self, n: int) -> bool:
        slowptr = n
        fastptr = self.getNext(n)
        while(fastptr!=1 and fastptr!=slowptr):
            slowptr = self.getNext(slowptr)
            fastptr = self.getNext(self.getNext(fastptr))
        return fastptr==1